scripts/list_apis.py

```diff
@@ -0,0 +1,277 @@
+        """
+        Return whether 'obj' is relevant to 'relevant_module'. If 'obj' is not a
+        module, then get its module from the __module__ attribute. Then check if
+        the module is part of the relevant_module by comparing their __file__ paths.
+
+        We don't want to crawl or list attributes of system modules or other
+        third-party modules that have been exposed via imports. Since submodules of
+        torch all live in the same directory, the heuristic here checks that the
+        module's __file__ contains /torch/.
+        """
+        # Don't look at this object if it is somewhere earlier in its own path
+        for item in path:
+            if id(obj) == id(item.obj):
+                # We've hit a circular include (i.e torch.__config__.torch.__config...)
+                # so give up
+                return True
+
+        module = obj
+        # print(obj, path)
+
+        if not isinstance(obj, ModuleType):
+            # If the object isn't already a module, resolve it to one
+            module_name = getattr(obj, "__module__", None)
+
+            if module_name is None:
+                return False
+
+            try:
+                module = importlib.import_module(module_name)
+            except ModuleNotFoundError as e:
+                # Likely we hit an incorrectly named module defined in C (so
+                # the name given to the CPython API isn't actually importable)
+                return True
+
+
+        if module == self.module:
+            # If the object is a direct child of the concered module, then don't
+            # skip
+            return False
+
+        if module in BUILTIN_MODULES:
+            # Skip built ins
+            return True
+
+        module_file = None
+        try:
+            # For some objects merely accessing an attribute causes an error, so catch
+            # that here
+            module_file = getattr(module, "__file__", None)
+        except RuntimeError as e:
+            return False
+
+        if not isinstance(module_file, str):
+            return False
+
+        # Check if the module is a submodule of the relevant module via checking
+        # their file paths
+        parent = Path(self.module.__file__).resolve().parent
+        maybe_child = Path(module_file).resolve()
+
+        return parent not in maybe_child.parents
+
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Recursively list all reachable objects in a Python module (library)."
+        + " This is used in PyTorch releases to determine the API surface changes"
+    )
+    parser.add_argument("--module", help="module to crawl", required=True)
+
+    def add_flag(name: str, default: bool, help: str) -> None:
+        parser.add_argument(f"--{name}", dest=name, help=help, action="store_true")
+        parser.add_argument(f"--no-{name}", dest=name, help=help, action="store_false")
+        parser.set_defaults(**{name: default})
```

Comment:
I don't quite understand this: in `main` above, you use this for `public`, `private`, and `errors`, but in `main` above, you implicitly coerce all three of those to booleans before using them, so the `--no-` variant seems to not do anything; could you clarify why you add it?

Reply:
oh never mind, sorry; I didn't see that you set the `default` for these when you call `add_flag` below